lib/crypto/ed25519/ed25519.go

```diff
@@ -44,6 +44,22 @@ type PublicKey ed25519.PublicKey
 // PublicKeyBytes is an encoded ed25519 public key
 type PublicKeyBytes [PublicKeyLength]byte
 
+// VerifySignature verifies a signature given a public key and a message
+func VerifySignature(publicKey, signature, message []byte) error {
+	pubKey, err := NewPublicKey(publicKey)
+	if err != nil {
+		return fmt.Errorf("ed25519: %w", err)
+	}
+
+	ok, err := pubKey.Verify(message, signature)
+	if err != nil || !ok {
```

Comment:
My suggestion would be:
```go
ok, err := pubKey.Verify(message, signature)
if err != nil {
  return fmt.Errorf("sr25519: %w: because %s", crypto.ErrSignatureVerificationFailed, err)
}

if !ok {
  return fmt.Errorf("sr25519: %w: for message 0x%x, signature 0x%x and public key 0x%x",
			crypto.ErrSignatureVerificationFailed, message, signature, publicKey)
}
```
since when `err != nil`, the cause of it will be explicit on `err`, e.g if one of those happen:
```go
if k.key == nil {
  return false, errors.New("nil public key")
}

if len(sig) != SignatureLength {
  return false, errors.New("invalid signature length")
}
```

and in the case where `err == nill && !ok`, the problem was really the the verification step, i.e:
```go
t := sr25519.NewSigningContext(SigningContext, msg)
return k.key.Verify(s, t), nil
```
The same would apply for `ed25518`.

Reply:
Sure, although don't wrap both errors with `crypto.ErrSignatureVerificationFailed` that looks strange. The parent calling function should wrap the error in this case. Instead you can just have

```go
ok, err := pubKey.Verify(message, signature)
if err != nil {
  return fmt.Errorf("sr25519: %w", err)
} else if !ok {
  return fmt.Errorf("sr25519: %w: for message 0x%x, signature 0x%x and public key 0x%x",
			crypto.ErrSignatureVerificationFailed, message, signature, publicKey)
}
```

?

Reply:
looking now it really seems strange that way, changed it on 1ee5e82 4012a66